### PR TITLE
les: fix light fetcher database race

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR fixes https://github.com/ethereum/go-ethereum/issues/16101 by avoiding a potential chain db data race that can cause checkAnnouncedHeaders being called with a nil header. There is also another check for missing parent of validated header (with an error log message) which in theory should not happen but still it would be nice to know about if it happens anyway in some weird corner case.

### CODE REVIEW

**Comment by holiman** [comment]
On mobile, so maybe I missed something, but it looks wrong. You check if header is nil... And then use header.Number in the error message. That'll blow up.

File: les/fetcher.go [hunk 1]
```diff
@@ -562,6 +562,10 @@ func (f *lightFetcher) checkAnnouncedHeaders(fp *fetcherPeerInfo, headers []*typ
 			// we ran out of recently delivered headers but have not reached a node known by this peer yet, continue matching
 			td = f.chain.GetTd(header.ParentHash, header.Number.Uint64()-1)
 			header = f.chain.GetHeader(header.ParentHash, header.Number.Uint64()-1)
+			if header == nil || td == nil {
+				log.Error("Missing parent of validated header", "hash", header.ParentHash, "number", header.Number.Uint64()-1)
```
Comment (holiman) [inline les/fetcher.go]: This will cause error if `header` is indeed `nil`

File: les/fetcher.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ type lightFetcher struct {
 	peers           map[*peer]*fetcherPeerInfo
 	lastUpdateStats *updateStatsEntry
 
-	lock       sync.Mutex // qwerqwerqwe
+	lock       sync.Mutex
```
Comment (holiman) [inline les/fetcher.go]: Couldn't you replace `qwerqwerqwe` with some documentation about specifically what asset this lock protects?

Comment (zsfelfoldi) [reply les/fetcher.go]: done

File: les/fetcher.go (hunk 1 above)
Comment (zsfelfoldi) [reply les/fetcher.go]: Right, this one was a really stupid mistake.

**Pull request merged**